### ISSUE TITLE
SAFECode RewriteOOB pass does not properly replace SSA values

### ISSUE THREAD

**Issue by llvmbot** [llvm/llvm-project#16868]
|  |  |
| --- | --- |
| Bugzilla Link | [16494](https://llvm.org/bz16494) |
| Version | unspecified |
| OS | All |
| Reporter | LLVM Bugzilla Contributor |

## Extended Description 
The SAFECode RewriteOOB pass attempts to replace the use of a checked pointer with the result of a run-time check (e.g., a bounds check).  However, it does not do so correctly, and the fact that it does not generates false positive errors in GNU Grep 2.14.

There are several issues with the existing code.  First, not all uses are replaced as the code only replaces uses that the check dominates.  There are other uses that the check may not dominate but should still see the rewritten value if control-flow passes through the check (e.g., a phi node that takes the checked pointer as an operand).

Another potential issue is that checks are not visited in dominator-tree order, so if they are rewritten in the wrong order, a use may use the result of an earlier check instead of a later check.

In short, the approach to doing OOB pointer rewriting needs to be reexamined.

**Comment by Endilll** [comment]
SAFECode is archived.
